app/helpers/heatmap_helper.rb

```diff
@@ -71,25 +35,434 @@ def sample_taxons_dict(params)
+    taxon_ids = HeatmapHelper.top_taxons_details(
+      samples,
+      background_id,
+      num_results,
+      sort_by,
+      species_selected,
+      categories,
+      threshold_filters,
+      read_specificity,
+      include_phage,
+      min_reads
+    ).pluck('tax_id')
+
     taxon_ids -= removed_taxon_ids
 
-    samples_taxons_details(samples, taxon_ids, background_id, species_selected, threshold_filters)
+    HeatmapHelper.samples_taxons_details(
+      samples,
+      taxon_ids,
+      background_id,
+      species_selected,
+      threshold_filters
+    )
+  end
+
+  def self.top_taxons_details(
+    samples,
+    background_id,
+    num_results,
+    sort_by,
+    species_selected,
+    categories,
+    threshold_filters = {},
+    read_specificity = READ_SPECIFICITY,
+    include_phage = INCLUDE_PHAGE,
+    min_reads = MINIMUM_READ_THRESHOLD
+  )
+    # return top taxons
+    results_by_pr = fetch_top_taxons(
+      samples,
+      background_id,
+      categories,
+      read_specificity,
+      include_phage,
+      num_results,
+      min_reads,
+      sort_by
+    )
+
+    sort = ReportHelper.decode_sort_by(sort_by)
+    count_type = sort[:count_type]
+    metric = sort[:metric]
+    candidate_taxons = {}
+    results_by_pr.each do |_pr_id, res|
+      pr = res["pr"]
+      taxon_counts = res["taxon_counts"]
+      sample_id = pr.sample_id
+
+      tax_2d = ReportHelper.taxon_counts_cleanup(taxon_counts)
+      HeatmapHelper.only_species_or_genus_counts!(tax_2d, species_selected)
+
+      rows = []
+      tax_2d.each do |_tax_id, tax_info|
+        rows << tax_info
+      end
+
+      HeatmapHelper.compute_aggregate_scores_v2!(rows)
+      rows = rows.select do |row|
+        row["NT"]["maxzscore"] >= MINIMUM_ZSCORE_THRESHOLD &&
+          # Note: these are applied *after* SQL filters, so results may not be
+          # 100% as expected .
+          HeatmapHelper.check_custom_filters(row, threshold_filters)
+      end
+
+      # Get the top N for each sample. This re-sorts on the same metric as in
+      # fetch_top_taxons SQL. We sort there first for performance.
+      rows.sort_by! { |tax_info| ((tax_info[count_type] || {})[metric] || 0.0) * -1.0 }
+      count = 1
+      rows.each do |row|
+        taxon = if candidate_taxons[row["tax_id"]]
+                  candidate_taxons[row["tax_id"]]
+                else
+                  { "tax_id" => row["tax_id"], "samples" => {} }
+                end
+        taxon["max_aggregate_score"] = row[sort[:count_type]][sort[:metric]] if
+          taxon["max_aggregate_score"].to_f < row[sort[:count_type]][sort[:metric]].to_f
+        taxon["samples"][sample_id] = [count, row["tax_level"], row["NT"]["zscore"], row["NR"]["zscore"]]
+        candidate_taxons[row["tax_id"]] = taxon
+        break if count >= num_results
```

Comment:
the original sorting and ranking behavior is preserved and applied on top of the SQL sort and rank so that results are guaranteed to be the same as before